runconfig/hostconfig.go

```diff
@@ -260,52 +260,44 @@ func NewCapList(caps []string) *CapList {
+	CPUShares            int64            `json:"CpuShares"` // CPU shares (relative weight vs. other containers)
+	CPUPeriod            int64            `json:"CpuPeriod"` // CPU CFS (Completely Fair Scheduler) period
+	CpusetCpus           string           // CpusetCpus 0-2, 0,1
+	CpusetMems           string           // CpusetMems 0-2, 0,1
+	CPUQuota             int64            `json:"CpuQuota"` // CPU CFS (Completely Fair Scheduler) quota
+	BlkioWeight          int64            // Block IO weight (relative weight vs. other containers)
+	OomKillDisable       bool             // Whether to disable OOM Killer or not
+	MemorySwappiness     *int64           // Tuning container memory swappiness behaviour
+	Privileged           bool             // Is the container in privileged mode
+	PortBindings         nat.PortMap      // Port mapping between the exposed port (container) and the host
+	Links                []string         // List of links (in the name:alias form)
+	PublishAllPorts      bool             // Should docker publish all exposed port for the container
+	DNS                  []string         `json:"Dns"`       // List of DNS server to lookup
+	DNSSearch            []string         `json:"DnsSearch"` // List of DNSSearch to look for
+	ExtraHosts           []string         // List of extra hosts
+	VolumesFrom          []string         // List of volumes to take from other container
+	Devices              []DeviceMapping  // List of devices to map inside the container
+	NetworkMode          NetworkMode      // Network namespace to use for the container
+	IpcMode              IpcMode          // IPC namespace to use for the container
+	PidMode              PidMode          // PID namespace to use for the container
+	UTSMode              UTSMode          // UTS namespace to use for the container
+	CapAdd               *CapList         // List of kernel capabilities to add to the container
+	CapDrop              *CapList         // List of kernel capabilities to remove from the container
+	GroupAdd             []string         // List of additional groups that the container process will run as
+	RestartPolicy        RestartPolicy    // Restart policy to be used for the container
+	SecurityOpt          []string         // List of string values to customize labels for MLS systems, such as SELinux.
+	ReadonlyRootfs       bool             // Is the container root filesystem in read-only
+	Ulimits              []*ulimit.Ulimit // List of ulimits to be set in the container
+	LogConfig            LogConfig        // Configuration of the logs for this container
+	CgroupParent         string           // Parent cgroup.
+	ConsoleSize          [2]int           // Initial console size on Windows
+	ExplicitExposedPorts bool             // True if --expose was passed from CLI
+	ExplicitPublishPorts bool             // True if --publish was passed from CLI
```

Comment:
Yes. I understand & appreciate the intention to have least impact to the existing code and hence you added the flag that will exactly mimic the client validation on the server side :-)
But, if you look at the logic and the intention behind that check, it will make it clear.
`if vals.netMode.IsContainer() && (vals.flPublish.Len() > 0 || *vals.flPublishAll == true) {`
`if vals.netMode.IsContainer() && vals.flExpose.Len() > 0 {`

If the network mode is `container:` which means it shares the network namespace with another container, any additional port-mapping or port-expose is not permitted. 
Hence checking for `config.ExposedPorts` and `hostconfig.PortMappings` will give you the exact same expected behavior. Since we are moving the validation from client to server side, I think it is appropriate to do that without having to introduce 2 additional flags just for a single purpose. Wont you agree ?